FORMS.md

```diff
@@ -0,0 +1,243 @@
+
+### States
+
+Many of these states are assigned or applicable to the `<input>` element. However, those Input states may be better reflected at the Field level. For example: dimming Field.Label for `disabled` elements; showing Field.Error only when an Input is `invalid`.
+
+- `active`: When the Input is activated by the user.
+- `autofill`: When an Input has been auto-filled by the browser.
+- `blank`: When an Input is empty.
+- `checked`: When Inputs like Radio and Checkbox are checked.
+- `dir`: The reading direction currently applied.
+- `disabled`: When an Input is disabled.
+- `enabled`: When an Input is enabled (rather than disabled). This is the default state.
+- `focus-within` When a child has `focus`.
+- `focus`: When the Input has focus.
+- `hover`: When the Input is hovered by a pointing device.
+- `in-range` When an Input is within the range constrictions. e.g. a slider.
+- `indeterminate`: When Inputs are in an indeterminate state. e.g. A "parent" checkbox for a group of mixed checked/unchecked Inputs.
+- `invalid`: When an Input has invalid contents/values.
+- `optional`: When an Input is optional.
+- `out-of-range`: When an Input is outside of range constrictions.
+- `placeholder-shown`: When placeholder text is visible.
+- `read-only`: When an Input cannot be changed by the user.
+- `read-write`: When an Input is user-editable (rather than read-only). This is the default state.
+- `required`: When an Input is required.
+- `target-within`: When a child is a `target`.
+- `target`: When the component is the target of a URL. e.g. `site.com/form#firstname`
+- `user-invalid`: When an Input is invalid, but only after a user has interacted with it.
+- `valid`: When an Input has valid contents/values.
+
+### Content
+
+- FieldLabel: a text label for the field (element may be `<label>` or `<legend>` depending on Field input)
+  - The Label also indicates the Field's optional/required status
+- FieldHint: context for the field
+- Input: the UI control
+- FieldValidation: any error or success state messaging for the field
+
+### Psuedo-structure
+
+```html
+// via props
+
+<Field
+  label="A field"
+  hint="This is a field you should fill in."
+  type="text"
+  name="a-field"
+  optional
+  disabled
+  autocomplete="fname"
+/>
+
+
+// via composition
+
+<Field>
+  <Field.Label>A field</Field.Label>
+  <Field.Hint>This is a field you should fill in.</Field.Hint>
+  <Input/>
+  <Field.Validation type="error">This is an error message.</Field.Validation>
+</Field>
+```
+
+### HTML
+
+```html
+<fieldset class="ods-field">
+  <label class="ods-field--label" for="first-name">
+    First name
+    <span class="ods-field--optional">Optional</span>
+  </label>
+  <aside class="ods-field--hint" id="first-name-hint">
+    Please enter your first name.
+  </aside>
+  <input class="ods-text-input" type="text" name="first-name" id="first-name" spellcheck="false" value="" autocomplete="fname" aria-describedby="first-name-hint first-name-error">
```

Comment:
have we considered restructuring such that `label` is an ancestor to the `input` element? As I'm sure you know this still allows the two elements to be associated correctly without the need for creating and assigning a unique ID at run time.

Reply:
Three hurdles make a `label`-parent solution problematic:

1. Screen readers are not uniform in how they treat `label > input` experiences. Some treat _all_ content of `label` as label text, which leads to confusing read-alouds. Some, sadly, do not follow the spec and fail to associate parent `label`s with their child `input`s.
2. `label`/`input` id associations are still required for other aria functionalities, like associating Hints and Errors with their appropriate fields.
3. Not all Field types allow for this construction. For example, rather than swapping `label` to `legend`, RadioGroup would require a different structure altogether than Text or Password because `legend` cannot wrap the entire `fieldset`. This becomes amplified when we get into custom Fields like Multi-Select or File Upload.